ui.py

```diff
@@ -1,75 +1,89 @@
+    def current_completion(self) -> Optional[CopilotPayloadCompletion]:
+        """The current chosen `completion`."""
+        completions = self.completions
+        return completions[self.completion_index] if completions else None
+
+    def choose_previous_completion(self) -> None:
+        """Set `completion_index` to be for the previous completion."""
+        self._set_completion_index(
+            self.completion_index - 1,
+            do_clamp=not self.view.settings().get("auto_complete_cycle", False),
+        )
 
-        return raw_display_text[:index] if following_text and index != -1 else raw_display_text
+    def choose_next_completion(self) -> None:
+        """Set `completion_index` to be for the next completion."""
+        self._set_completion_index(
+            self.completion_index + 1,
+            do_clamp=not self.view.settings().get("auto_complete_cycle", False),
+        )
 
     def hide(self) -> None:
-        set_copilot_view_setting(self.view, "is_visible", False)
-
-        PopupCompletion.hide(self.view)
+        """Hide Copilot's completion popup."""
+        # to prevent from hiding other plugin's popup
+        if self.is_visible:
+            _PopupCompletion.hide(self.view)
 
-    def show(self, region: Tuple[int, int], completions: List[CopilotPayloadCompletion], cycle: int = 0) -> None:
-        if not completions:
+    def show(self) -> None:
+        """Show Copilot's completion popup."""
+        completion = self.current_completion
+        if not completion:
             return
 
-        completion = completions[cycle % len(completions)]
-
-        display_text = self.get_display_text(region, completion["displayText"])
-        if not display_text:
+        # the text after completion is the same
+        current_line = self.view.line(completion["positionSt"])
+        if completion["text"] == self.view.substr(current_line):
             return
 
-        set_copilot_view_setting(self.view, "is_visible", True)
-        set_copilot_view_setting(self.view, "region", region)
-        set_copilot_view_setting(self.view, "uuid", completion["uuid"])
-        set_copilot_view_setting(self.view, "text", completion["text"])
-        set_copilot_view_setting(self.view, "display_text", display_text)
+        _PopupCompletion(self.view).show()
+
+    def _set_completion_index(self, value: int, do_clamp: bool = False) -> None:
+        """
+        Set `completion_index`.
 
-        PopupCompletion(self.view, region, display_text).show()
+        :param      value:     The wanted new completion index.
+        :param      do_clamp:  Should the `value` be clamped if it's
+                               out-of-bound? Otherwise, treat it as cyclic.
+        """
+        completions_cnt = len(self.completions)
+        if completions_cnt:
+            value = clamp(value, 0, completions_cnt - 1) if do_clamp else value % completions_cnt
+        else:
+            value = 0
```

Comment:
I think the `completion_index` should be always within the range, so `do_clamp` should be always on.

To simplify the implementation we could still use `%`.
But when the `value` is negative we simply use `completions_cnt + value` as the final index before the `%` thing.  
This way the index will be always within the range

Reply:
There is a reason why ST introduce `auto_complete_cycle` setting.

Reply:
Hm, I wasn't aware of `auto_complete_cycle`. To be honest, I've been looking for this setting and even thought there was a bug in Sublime 🙂
Yes, it makes sense to respect this setting then 👍🏻